### DISCUSSION TITLE
Make the inclusion of heap data necessary for profliling optional?

### DISCUSSION THREAD

**Discussion by zakkak** [oracle/graal#7707] [Q&A] [answered]
Hello,

I was looking into the changes that went in between GraalVM for JDK 17 (23.0) and GraalVM for JDK 21 (23.1) to see what is causing the binary size increase we are observing when using GraalVM for JDK 21 and the culprit is that the main contributors are:

1. [**Add support for profiling of topmost frame**](https://github.com/oracle/graal/pull/7003) which adds ~1MB (~1.75% of the total image size) of data to the native image of [quarkus-startstop](https://github.com/quarkus-qe/quarkus-startstop/).
2. [**ProfilingSampler does not need local variable values**](https://github.com/oracle/graal/pull/6763) (specifically [the commit "Always store bci in frame info"](https://github.com/oracle/graal/commit/d747c30c7691012c39989a8597fd850c68b740ad)) which adds ~1.7MB (~3% of the total image size) of data to the native image of [quarkus-startstop](https://github.com/quarkus-qe/quarkus-startstop/).

To my understanding both changes are necessary to improve the accuracy of the ~async-profiler (and possibly other tools as well)~ async-sampler which is used by third party profilers. Is that right?

Would it make sense to introduce a new option that would control whether such additional data would be included or not in the image? I was thinking something like `-perf`, similarly to the `-gperf` proposed in https://github.com/oracle/graal/issues/7370.

Thank you

cc @christianwimmer @jovanstevanovic

**Comment by jovanstevanovic** [comment, marked as answer]
Hi @zakkak, for bullet one, yes, it improves async-sampler results. Additionally, we are using it in SubstrateDiagnostics and ThreadPrinter. For example, now each SEGFAULT has at least a method and class for a top frame.

We had this debate about whether it is necessary or not, and the conclusion was that the benefits are higher than the costs.
